server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,478 @@
+        int activeMerges = (int) (submittedMergesCount - doneMergesCount);
+        if (activeMerges > configuredMaxMergeCount && shouldThrottleIncomingMerges.get() == false) {
+            // maybe enable merge task throttling
+            synchronized (shouldThrottleIncomingMerges) {
+                if (shouldThrottleIncomingMerges.getAndSet(true) == false) {
+                    enableIndexingThrottling(executingMergesCount, activeMerges - executingMergesCount, configuredMaxMergeCount);
+                }
+            }
+        } else if (activeMerges <= configuredMaxMergeCount && shouldThrottleIncomingMerges.get()) {
+            // maybe disable merge task throttling
+            synchronized (shouldThrottleIncomingMerges) {
+                if (shouldThrottleIncomingMerges.getAndSet(false)) {
+                    disableIndexingThrottling(executingMergesCount, activeMerges - executingMergesCount, configuredMaxMergeCount);
+                }
+            }
+        }
+    }
+
+    // synchronized so that {@code #closed}, {@code #currentlyRunningMergeTasks} and {@code #backloggedMergeTasks} are modified atomically
+    private synchronized boolean runNowOrBacklog(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        if (closed) {
+            // Do not backlog or execute tasks when closing the merge scheduler, instead abort them.
+            mergeTask.abortOnGoingMerge();
+            throw new ElasticsearchException("merge task aborted because scheduler is shutting down");
+        }
+        if (currentlyRunningMergeTasks.size() < config.getMaxThreadCount()) {
+            boolean added = currentlyRunningMergeTasks.put(mergeTask.onGoingMerge.getMerge(), mergeTask) == null;
+            assert added : "starting merge task [" + mergeTask + "] registered as already running";
+            return true;
+        } else {
+            backloggedMergeTasks.add(mergeTask);
+            return false;
+        }
+    }
+
+    private void mergeDone(MergeTask mergeTask) {
+        try {
+            synchronized (this) {
+                boolean removed = currentlyRunningMergeTasks.remove(mergeTask.onGoingMerge.getMerge()) != null;
+                assert removed : "completed merge task [" + mergeTask + "] not registered as running";
+                // when one merge is done, maybe a backlogged one can now execute
+                enqueueBackloggedTasks();
+                // signal here, because, when closing, we wait for all currently running merges to finish
+                maybeSignalAllMergesDoneAfterClose();
+            }
+        } finally {
+            doneMergeTaskCount.incrementAndGet();
+            checkMergeTaskThrottling();
+        }
+    }
+
+    private synchronized void maybeSignalAllMergesDoneAfterClose() {
+        if (closed && currentlyRunningMergeTasks.isEmpty() && closedWithNoCurrentlyRunningMerges.getCount() > 0) {
```

Comment:
@henningandersen Re what you brought up about the `MergeScheduler` closing promptly while there are enqueued merge tasks to execute. This code here waits only for already running merges to finish (the same behavior as the `ConcurrentMergeScheduler`), the enqueued ones are completely ignoed. When enqueued ones eventually get to the point of executing, they will be promptly aborted in `ThreadPoolMergeScheduler#runNowOrBacklog`.

Reply:
There's now also a test for that `ThreadPoolMergeSchedulerTests#testSchedulerCloseWaitsForRunningMerge`.